NPB/CG/cg.cpp

```diff
@@ -0,0 +1,1052 @@
+	firstcol = 1;
+	lastcol  = NA;
+
+	if (NA == 1400 && NONZER == 7 && NITER == 15 && SHIFT == 10.0) {
+		class_npb = 'S';
+		zeta_verify_value = 8.5971775078648;
+	} else if (NA == 7000 && NONZER == 8 && NITER == 15 && SHIFT == 12.0) {
+		class_npb = 'W';
+		zeta_verify_value = 10.362595087124;
+	} else if (NA == 14000 && NONZER == 11 && NITER == 15 && SHIFT == 20.0) {
+		class_npb = 'A';
+		zeta_verify_value = 17.130235054029;
+	} else if (NA == 75000 && NONZER == 13 && NITER == 75 && SHIFT == 60.0) {
+		class_npb = 'B';
+		zeta_verify_value = 22.712745482631;
+	} else if (NA == 150000 && NONZER == 15 && NITER == 75 && SHIFT == 110.0) {
+		class_npb = 'C';
+		zeta_verify_value = 28.973605592845;
+	} else {
+		class_npb = 'U';
+	}
+
+	if (0 == dash::myid()) {
+
+		printf("\n\n NAS Parallel Benchmarks 4.0 C++ DASH version"" - CG Benchmark\n");
+		printf("\n\n Developed by: Dalvan Griebler <dalvan.griebler@acad.pucrs.br>\n");
+		printf("\n\n DASH version by: Nicco Mietzsch <nicco.mietzsch@campus.lmu.de>\n");
+		printf(" Size: %10d\n", NA);
+		printf(" Iterations: %5d\n", NITER);
+
+
+		naa = NA;
+		nzz = NZ;
+
+		/*--------------------------------------------------------------------
+		c  Initialize random number generator
+		c-------------------------------------------------------------------*/
+		tran    = 314159265.0;
+		amult   = 1220703125.0;
+		zeta    = randlc( &tran, amult );
+
+		/*--------------------------------------------------------------------
+		c
+		c-------------------------------------------------------------------*/
+		static int colidx_init[NZ+1];	/* colidx[1:NZ] */
+		static int rowstr_init[NA+1+1];	/* rowstr[1:NA+1] */
+
+		static double a_init[NZ+1];		/* a[1:NZ] */
+
+
+
+		static int iv[2*NA+1+1];	/* iv[1:2*NA+1] */
+		static int arow[NZ+1];		/* arow[1:NZ] */
+		static int acol[NZ+1];		/* acol[1:NZ] */
+
+		static double v[NA+1+1];	/* v[1:NA+1] */
+		static double aelt[NZ+1];	/* aelt[1:NZ] */
+
+
+		makea(naa, nzz, a_init, colidx_init, rowstr_init, NONZER,
+		firstrow, lastrow, firstcol, lastcol,
+		RCOND, arow, acol, aelt, v, iv, SHIFT);
+
+		for(i = 0; i < NZ+1; ++i) colidx_global[i] = colidx_init[i];
+		for(i = 0; i < NA+1+1; ++i) rowstr_global[i] = rowstr_init[i];
+		for(i = 0; i < NZ+1; ++i) a_global[i] = a_init[i];
+
+
+		/*--------------------------------------------------------------------
+		c  set starting vector to (1, 1, .... 1)
+		c-------------------------------------------------------------------*/
+
+		for (i = 1; i <= NA; i++) {
+			x[i] = 1.0;
+		}
+
+	}
+	dash::barrier();
+
+	//setup local arrays
+	// first, setup rowstr
+
+	int elem_per_unit = ceil(((double) NA+1)/dash::size());
+
+	dash::Array<int> rowstr((elem_per_unit+1)*dash::size());
+
+	for(int i = 0; i < elem_per_unit+1; i++) {
+			rowstr.local[i] = rowstr_global[min(dash::myid()*elem_per_unit + i, NA+1)];
+	}
+
+
+	//now setup a and colidx
+
+	typedef dash::CSRPattern<1>           pattern_t;
+  typedef int                           index_t;

```

Comment:
Could you please `typedef` the pattern (and also the `dash::Matrix` using it) somewhere centrally so that it is easy to change? 

Also: is it guranteed that the matrix will never have more than 2G elements?

Reply:
and instead of "typedef" i would prefer "using"